### PR TITLE
fix(amplify-category-function): add length validation for secret value

### LINKED ISSUE #8304: Add input validation to function secret value prompt
### Before opening, please confirm:

- [X] I have installed the latest version of the Amplify CLI (see above), and confirmed that the issue still persists.
- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-cli/issues?q=is%3Aissue+).
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-cli/blob/master/CONTRIBUTING.md#bug-reports).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

### How did you install the Amplify CLI?

yarn

### If applicable, what version of Node.js are you using?

v16.9.0

### Amplify CLI Version

6.1.0

### What operating system are you using?

Mac

### Amplify Categories

function

### Amplify Commands

add

### Describe the bug

Add input validation for function secrets' value, disallowing developer from continuing before adding a valid value. By doing so preventing the following error from appearing after completing secrets prompts **when we provide no value**. Error states there was an error adding the function resource due to the error, however function is successfully added and secret `mysecret` is not created. 

```shell
> amplify add function
⚠️ Amplify command hooks are disabled in the current execution environment.
⚠️ See https://docs.amplify.aws/cli/usage/command-hooks/ for more information.
? Select which capability you want to add: Lambda function (serverless function)
? Provide an AWS Lambda function name: myfunction
? Choose the runtime that you want to use: NodeJS
? Choose the function template that you want to use: Hello World

Available advanced settings:
- Resource access permissions
- Scheduled recurring invocation
- Lambda layers configuration
- Environment variables configuration
- Secret values configuration

? Do you want to configure advanced settings? Yes
? Do you want to access other resources in this project from your Lambda function? No
? Do you want to invoke this function on a recurring schedule? No
? Do you want to enable Lambda layers for this function? No
? Do you want to configure environment variables for this function? No
? Do you want to configure secret values this function can access? Yes
? Enter a secret name (this is the key used to look up the secret value): mysecret
? Enter the value for mysecret: [hidden]
? What do you want to do? I'm done
ValidationException: 1 validation error detected: Value at 'value' failed to satisfy constraint: Member must have length greater than or equal to 1.
    at Request.extractError (/Users/josef/.config/yarn/global/node_modules/aws-sdk/lib/protocol/json.js:52:27)
    at Request.callListeners (/Users/josef/.config/yarn/global/node_modules/aws-sdk/lib/sequential_executor.js:106:20)
    at Request.emit (/Users/josef/.config/yarn/global/node_modules/aws-sdk/lib/sequential_executor.js:78:10)
    at Request.emit (/Users/josef/.config/yarn/global/node_modules/aws-sdk/lib/request.js:688:14)
    at Request.transition (/Users/josef/.config/yarn/global/node_modules/aws-sdk/lib/request.js:22:10)
    at AcceptorStateMachine.runTo (/Users/josef/.config/yarn/global/node_modules/aws-sdk/lib/state_machine.js:14:12)
    at /Users/josef/.config/yarn/global/node_modules/aws-sdk/lib/state_machine.js:26:10
    at Request.<anonymous> (/Users/josef/.config/yarn/global/node_modules/aws-sdk/lib/request.js:38:9)
    at Request.<anonymous> (/Users/josef/.config/yarn/global/node_modules/aws-sdk/lib/request.js:690:12)
    at Request.callListeners (/Users/josef/.config/yarn/global/node_modules/aws-sdk/lib/sequential_executor.js:116:18)
    at Request.emit (/Users/josef/.config/yarn/global/node_modules/aws-sdk/lib/sequential_executor.js:78:10)
    at Request.emit (/Users/josef/.config/yarn/global/node_modules/aws-sdk/lib/request.js:688:14)
    at Request.transition (/Users/josef/.config/yarn/global/node_modules/aws-sdk/lib/request.js:22:10)
    at AcceptorStateMachine.runTo (/Users/josef/.config/yarn/global/node_modules/aws-sdk/lib/state_machine.js:14:12)
    at /Users/josef/.config/yarn/global/node_modules/aws-sdk/lib/state_machine.js:26:10
    at Request.<anonymous> (/Users/josef/.config/yarn/global/node_modules/aws-sdk/lib/request.js:38:9)
    at Request.<anonymous> (/Users/josef/.config/yarn/global/node_modules/aws-sdk/lib/request.js:690:12)
    at Request.callListeners (/Users/josef/.config/yarn/global/node_modules/aws-sdk/lib/sequential_executor.js:116:18)
    at callNextListener (/Users/josef/.config/yarn/global/node_modules/aws-sdk/lib/sequential_executor.js:96:12)
    at IncomingMessage.onEnd (/Users/josef/.config/yarn/global/node_modules/aws-sdk/lib/event_listeners.js:336:13)
    at IncomingMessage.emit (node:events:406:35)
    at IncomingMessage.emit (node:domain:475:12)
    at endReadableNT (node:internal/streams/readable:1343:12)
    at processTicksAndRejections (node:internal/process/task_queues:83:21)
There was an error adding the function resource
```

Despite the above error, secrets are added to the function's CloudFormation template and `function-parameters.json` file, which has its own caveat as documented in the NOTE below.

- `function-parameters.json`
```json
{
  "lambdaLayers": [],
  "secretNames": [
    "mysecret"
  ]
}
```
- CloudFormation template
```json
{
  "AWSTemplateFormatVersion": "2010-09-09",
  "Description": "Lambda Function resource stack creation using Amplify CLI",
  "Parameters": {
    "CloudWatchRule": {
      "Type": "String",
      "Default": "NONE",
      "Description": " Schedule Expression"
    },
    "deploymentBucketName": {
      "Type": "String"
    },
    "env": {
      "Type": "String"
    },
    "s3Key": {
      "Type": "String"
    },
    "secretsPathAmplifyAppId": {
      "Type": "String"
    }
  },
...
"Environment": {
          "Variables": {
            "ENV": {
              "Ref": "env"
            },
            "REGION": {
              "Ref": "AWS::Region"
            },
            "mysecret": {
              "Fn::Join": [
                "",
                [
                  {
                    "Fn::Sub": [
                      "/amplify/${appId}/${env}/AMPLIFY_${functionName}_",
                      {
                        "appId": {
                          "Ref": "secretsPathAmplifyAppId"
                        },
                        "env": {
                          "Ref": "env"
                        },
                        "functionName": "myfunction"
                      }
                    ]
                  },
                  "mysecret"
                ]
              ]
```

---

**NOTE**: if we have another existing environment locally (not connected to git), since the failed secret's information is added to the function's CloudFormation template we are not prompted for the secret's value when switching envs which ultimately leads to a failed deployment:

```shell
> amplify env checkout dev
✔ Initialized provider successfully.
Initialized your environment successfully.

> amplify push -y
⚠️ Amplify command hooks are disabled in the current execution environment.
⚠️ See https://docs.amplify.aws/cli/usage/command-hooks/ for more information.
✔ Successfully pulled backend environment dev from the cloud.

    Current Environment: dev
    
┌──────────┬─────────────────┬───────────┬───────────────────┐
│ Category │ Resource name   │ Operation │ Provider plugin   │
├──────────┼─────────────────┼───────────┼───────────────────┤
│ Function │ myfunction      │ Update    │ awscloudformation │
└──────────┴─────────────────┴───────────┴───────────────────┘
⠇ Updating resources in the cloud. This may take a few minutes...

UPDATE_IN_PROGRESS amplify-secrets-dev-115438 AWS::CloudFormation::Stack Wed Sep 29 2021 12:17:23 GMT-0500 (Central Daylight Time) User Initiated
⠇ Updating resources in the cloud. This may take a few minutes...

UPDATE_IN_PROGRESS          functionmyfunction   AWS::CloudFormation::Stack Wed Sep 29 2021 12:17:29 GMT-0500 (Central Daylight Time)                                                                        
UPDATE_FAILED               functionmyfunction    AWS::CloudFormation::Stack Wed Sep 29 2021 12:17:30 GMT-0500 (Central Daylight Time) Parameters: [secretsPathAmplifyAppId] must have values                 
UPDATE_ROLLBACK_IN_PROGRESS amplify-secrets-dev-115438 AWS::CloudFormation::Stack Wed Sep 29 2021 12:17:31 GMT-0500 (Central Daylight Time) The following resource(s) failed to update: [functionmyfunction]. 
⠧ Updating resources in the cloud. This may take a few minutes...

UPDATE_COMPLETE                              functionmyfunction    AWS::CloudFormation::Stack Wed Sep 29 2021 12:17:35 GMT-0500 (Central Daylight Time) 
UPDATE_ROLLBACK_COMPLETE_CLEANUP_IN_PROGRESS amplify-secrets-dev-115438 AWS::CloudFormation::Stack Wed Sep 29 2021 12:17:36 GMT-0500 (Central Daylight Time) 
UPDATE_COMPLETE                              functionmyfunction    AWS::CloudFormation::Stack Wed Sep 29 2021 12:17:38 GMT-0500 (Central Daylight Time) 
UPDATE_ROLLBACK_COMPLETE                     amplify-secrets-dev-115438 AWS::CloudFormation::Stack Wed Sep 29 2021 12:17:38 GMT-0500 (Central Daylight Time) 
⠦ Updating resources in the cloud. This may take a few minutes...

Following resources failed

✖ An error occurred when pushing the resources to the cloud

Resource is not in the state stackUpdateComplete
An error occurred during the push operation: Resource is not in the state stackUpdateComplete
```

### Expected behavior

Do not allow developer to continue through secrets prompts without adding a valid value, or continue gracefully while letting the developer know that no valid value was provided and thus secret will not be created nor added to the function.

```shell
? Enter a secret name (this is the key used to look up the secret value): mysecret
? Enter the value for mysecret: [hidden]
⚠️ Invalid value was provided. Value must have a length greater than or equal to 1
? Enter the value for mysecret: [hidden] # myvalue
? What do you want to do? I'm done
```

OR

```shell
? Enter a secret name (this is the key used to look up the secret value): mysecret
? Enter the value for mysecret: [hidden]
⚠️ Invalid value was provided. Secret will not be created
? What do you want to do? I'm done
```

Both workflows should allow developer to continue without adding secrets information to the CloudFormation template until a valid value is provided.

### Reproduction steps

1. `amplify init -y`
2. `amplify add function`
3. add secret
4. provide no value for secret value prompt (hit enter when prompted)
5. choose "I'm done"
6. Observe error

### GraphQL schema(s)

<details>

```graphql
# Put schemas below this line


```

</details>

### Log output

<details>

```
# Put your logs below this line


```

</details>

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by SebasRod23**
#### Description of changes
Validate that the secret value provided by the user has a length between 1 and 2048 characters

#### Issue #, if available
Closes #8304

#### Description of how you validated changes
The cli now doesn't allow you to continue until you have provided a valid secret value, similar to how the validation of the secret name works.

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [ ] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/master/CONTRIBUTING.md#tests)
- [ ] Relevant documentation is changed or added (and PR referenced)
- [ ] New AWS SDK calls or CloudFormation actions have been added to relevant test and service IAM policies

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by danielleadams — CHANGES_REQUESTED** [review]
Hi @SebasRod23, thanks for the contribution. Do you mind adding a unit test? You can add a new file named `secretValuesWalkthrough.test.ts` to the `__tests__/provider-utils/awscloudformation/service-walkthroughs` directory.

**Comment by SebasRod23** [comment]
No problem, I can do that

File: packages/amplify-category-function/src/provider-utils/awscloudformation/service-walkthroughs/secretValuesWalkthrough.ts [hunk 1]
```diff
@@ -176,12 +176,20 @@ const enterSecretName = async (invalidNames: string[]) =>
 
 const secretValueDefaultMessage = (secretName: string) => `Enter the value for ${secretName}:`;
 
+export const secretValueValidator = () => (input?: string) => {
+  if (!input || input.length === 0 || input.length > 2048) {
```
Comment (cjihrig) [inline packages/amplify-category-function/src/provider-utils/awscloudformation/service-walkthroughs/secretValuesWalkthrough.ts:180]: Are the `!input` and `input.length === 0` checks both necessary?

**Review by cjihrig — COMMENTED** [review]
LGTM, with a small question.

File: packages/amplify-category-function/src/provider-utils/awscloudformation/service-walkthroughs/secretValuesWalkthrough.ts (hunk 1 above)
Comment (SebasRod23) [reply packages/amplify-category-function/src/provider-utils/awscloudformation/service-walkthroughs/secretValuesWalkthrough.ts:180]: I think `input.length === 0` should be enough, but I added `!input` just in case `input` got a value of `undefined`

Comment (cjihrig) [reply packages/amplify-category-function/src/provider-utils/awscloudformation/service-walkthroughs/secretValuesWalkthrough.ts:180]: Instead of `!input`, can you use `typeof input !== 'string'`.

Comment (SebasRod23) [reply packages/amplify-category-function/src/provider-utils/awscloudformation/service-walkthroughs/secretValuesWalkthrough.ts:180]: Sure, no problem

File: packages/amplify-category-function/src/provider-utils/awscloudformation/service-walkthroughs/secretValuesWalkthrough.ts [hunk 2]
```diff
@@ -176,12 +176,20 @@ const enterSecretName = async (invalidNames: string[]) =>
...
+  return true;
+};
+
 const enterSecretValue = async (message: string) =>
   (
     await inquirer.prompt<{ secretValue: string }>({
       type: 'password',
       name: 'secretValue',
       message,
+      validate: secretValueValidator(),
```
Comment (danielleadams) [inline packages/amplify-category-function/src/provider-utils/awscloudformation/service-walkthroughs/secretValuesWalkthrough.ts:192]: This looks good - I don't think you need to invoked `secretValueValidator` here. I think you can pass in the function without invoking it and removing the top level function from `secretValueValidator`.

File: packages/amplify-category-function/src/provider-utils/awscloudformation/service-walkthroughs/secretValuesWalkthrough.ts [hunk 3]
```diff
@@ -176,12 +176,20 @@ const enterSecretName = async (invalidNames: string[]) =>
 
 const secretValueDefaultMessage = (secretName: string) => `Enter the value for ${secretName}:`;
 
+export const secretValueValidator = () => (input?: string) => {
```
Comment (danielleadams) [inline packages/amplify-category-function/src/provider-utils/awscloudformation/service-walkthroughs/secretValuesWalkthrough.ts:179]: I think the `validate` parameter only needs the function, and doesn't need the top level one.
```suggestion
export const secretValueValidator = (input?: string) => {
```

Comment (SebasRod23) [reply packages/amplify-category-function/src/provider-utils/awscloudformation/service-walkthroughs/secretValuesWalkthrough.ts:179]: Sure, that seems better

**Review by danielleadams — APPROVED** [review]

**Review by jhockett — APPROVED** [review]

**Comment by khrisrichardson** [comment]
Per the [SecretsManager documentation](https://docs.aws.amazon.com/secretsmanager/latest/apireference/API_CreateSecret.html#API_CreateSecret_RequestParameters), the length constraints for both SecretBinary and SecretString are

```
Length Constraints: Minimum length of 0. Maximum length of 65536.
```

Per the [Systemsmanager documentation](https://docs.aws.amazon.com/systems-manager/latest/APIReference/API_PutParameter.html#API_PutParameter_RequestParameters), the length constraints for parameter values are as follows

```
Standard parameters have a value limit of 4 KB. Advanced parameters have a value limit of 8 KB.
```

In other words the limit of 2048 characters is too constrained.
